app/PaymentDrivers/BlockonomicsPaymentDriver.php

```diff
@@ -102,50 +102,163 @@ public function processWebhookRequest(PaymentWebhookRequest $request)
+
+        SystemLogger::dispatch(
+            [
+                'debug' => 'payment_invoice_relationship',
+                'payment_id' => $payment->id,
+                'payment_amount' => $payment->amount,
+                'payment_status' => $payment->status_id,
+                'invoice_id' => $invoice->id,
+                'invoice_balance_before' => $original_balance,
+                'invoice_amount' => $invoice->amount,
+                'payment_invoices_count' => $payment_invoices->count(),
+                'invoice_payments_count' => $invoice_payments->count(),
+                'invoice_paid_to_date' => $invoice->paid_to_date ?? 'null',
+            ],
+            SystemLog::CATEGORY_GATEWAY_RESPONSE,
+            SystemLog::EVENT_GATEWAY_SUCCESS,
+            SystemLog::TYPE_BLOCKONOMICS,
+            $payment->client ?? $company->clients()->first(),
+            $company,
+        );
+
+        // Ensure the payment is properly linked to the invoice
+        if (!$payment->invoices()->where('invoices.id', $invoice->id)->exists()) {
+            // Manually attach the payment to the invoice if not linked
+            $payment->invoices()->attach($invoice->id, [
+                'amount' => $payment->amount,
+                'refunded' => 0,
+            ]);
+        }
+
+        // Force InvoiceNinja to recalculate the invoice balance and amounts
+        $invoice = $invoice->calc()->getInvoice();
+
+        // Debug: Check if calc() worked properly
+        $total_payments = $invoice->payments()
+            ->where('status_id', Payment::STATUS_COMPLETED)
+            ->sum('payments.amount');
+
+        $expected_balance = $invoice->amount - $total_payments;
+
+        // If calc() didn't update the balance properly, do manual calculation
+        if ($invoice->balance != $expected_balance) {
+            SystemLogger::dispatch(
+                [
+                    'debug' => 'manual_balance_calculation',
+                    'invoice_balance_from_calc' => $invoice->balance,
+                    'expected_balance' => $expected_balance,
+                    'total_payments' => $total_payments,
+                    'invoice_amount' => $invoice->amount,
+                ],
+                SystemLog::CATEGORY_GATEWAY_RESPONSE,
+                SystemLog::EVENT_GATEWAY_SUCCESS,
+                SystemLog::TYPE_BLOCKONOMICS,
+                $payment->client ?? $company->clients()->first(),
+                $company,
+            );
+
+            // Manual calculation - this will handle negative balances correctly
+            $invoice->balance = $expected_balance;
+            $invoice->paid_to_date = $total_payments;
+        }
+
+        // Set correct status based on final balance
+        if ($invoice->balance <= 0) {
+            $invoice->status_id = Invoice::STATUS_PAID;
+        } else {
+            $invoice->status_id = Invoice::STATUS_PARTIAL;
+        }
+
+        $invoice->save();
```

Comment:
mutating the invoice state has consequences, you cannot mutate the invoice state like this.